commit 84e2674e937814b7b230a4aadc7edbb39158fda2 (refs/changes/36/2136/1)
Author: Marsell Kukuljevic <marsell@joyent.com>
Date:   2017-06-23T00:09:42+12:00 (2 years, 4 months ago)
    
    PUBAPI-1405: GetConfig fails if the account doesn't yet have a dclocalconfig

diff --git a/lib/config.js b/lib/config.js
index f9ef9e2..9407baf 100644
--- a/lib/config.js
+++ b/lib/config.js
@@ -133,7 +133,11 @@ function getConfigFromUFDS(req, callback) {
     req.sdc.ufds.getDcLocalConfig(account, dc,
             function _afterConfGet(err, conf) {
         if (err) {
-            return callback(translateErr(err));
+            if (err.name !== 'ResourceNotFoundError') {
+                return callback(translateErr(err));
+            }
+
+            conf = {}; // treat an empty hash as default
         }
 
         req.log.debug({ account: account, dc: dc, config: conf },
diff --git a/test/common.js b/test/common.js
index 7f5f1e9..128bf13 100644
--- a/test/common.js
+++ b/test/common.js
@@ -445,6 +445,41 @@ function waitForMahiCache(mahiclient, apath, cb) {
 }
 
 
+function waitForAccountConfigReady(client, cb) {
+    assert.object(client, 'client');
+    assert.func(cb, 'callback');
+
+    var nbTries = 0;
+    var MAX_NB_TRIES = 10;
+    var TRY_DELAY_IN_MS = 1000;
+
+    function getConfig() {
+        ++nbTries;
+        if (nbTries >= MAX_NB_TRIES) {
+            cb(new Error('max number of tries reached'));
+            return;
+        }
+
+        client.get('/my/config', function onGetConfig(err, req, res, config) {
+            if (err) {
+                cb(err);
+                return;
+            }
+
+            if (config.default_network) {
+                cb();
+            } else {
+                setTimeout(getConfig, TRY_DELAY_IN_MS);
+            }
+
+            return;
+        });
+    }
+
+    getConfig();
+}
+
+
 // Creates a temporary user, invokes bodyCb(), destroys the user, then invokes
 // cb(). Useful for running tests in bodyCb() with a user that'll be destroyed
 // after bodyCb() completes.
@@ -595,6 +630,12 @@ function setup(opts, cb) {
         },
         function setupPackage(_, next) {
             addPackage(userClient, SDC_128_PACKAGE, next);
+        },
+        function waitUserClientConfig(_, next) {
+            waitForAccountConfigReady(userClient, next);
+        },
+        function waitOtherUserClientConfig(_, next) {
+            waitForAccountConfigReady(otherUserClient, next);
         }
     ] }, function (err) {
         if (err) {
